src/rulesets/LinearAlgebra/factorization.jl

```diff
@@ -74,64 +75,66 @@ function frule(
+            factors1 = factors[1:q, :]
+            L2 = factors[(q + 1):end, :]
+            ∂A1 = ∂A[1:q, :]
+            ∂A2 = ∂A[(q + 1):end, :]
+            ∂U = triu(∂factors[1:q, :])
+        end
+        U = UpperTriangular(factors1)
+        L1 = UnitLowerTriangular(factors1)
+        tril!(lmul!(L1', ∂A1), -1)
+        ∂A1 .+= triu!(mul!(rmul!(∂U, U'), L2', ∂A2, -1, 1))
+        ldiv!(L1', ∂A1)
+        rdiv!(∂A, U')
+    end
+    if pivot isa LU_RowMaximum
+        ∂A = ∂A[invperm(F.p), :]
+    end
+    return NoTangent(), ∂A, NoTangent()
+end
+_lu_pullback(ΔF::AbstractThunk, A, pivot, F) = _lu_pullback(unthunk(ΔF), A, pivot, F)
 function rrule(
     ::typeof(lu), A::StridedMatrix, pivot::Union{LU_RowMaximum,LU_NoPivot}; kwargs...
 )
     F = lu(A, pivot; kwargs...)
-    function lu_pullback(ΔF::Tangent)
-        Δfactors = ΔF.factors
-        Δfactors isa AbstractZero && return (NoTangent(), Δfactors, NoTangent())
-        factors = F.factors
-        ∂factors = eltype(A) <: Real ? real(Δfactors) : Δfactors
-        ∂A = similar(factors)
-        m, n = size(A)
-        q = min(m, n)
-        if m == n  # square A
-            # ∂A = P' * (L' \ (tril(L' * ∂L, -1) + triu(∂U * U')) / U')
-            L = UnitLowerTriangular(factors)
-            U = UpperTriangular(factors)
-            ∂U = UpperTriangular(∂factors)
-            tril!(copyto!(∂A, ∂factors), -1)
-            lmul!(L', ∂A)
-            copyto!(UpperTriangular(∂A), UpperTriangular(∂U * U'))
-            rdiv!(∂A, U')
-            ldiv!(L', ∂A)
-        elseif m < n  # wide A, system is [P*A1 P*A2] = [L*U1 L*U2]
-            triu!(copyto!(∂A, ∂factors))
-            @views begin
-                factors1 = factors[:, 1:q]
-                U2 = factors[:, (q + 1):end]
-                ∂A1 = ∂A[:, 1:q]
-                ∂A2 = ∂A[:, (q + 1):end]
-                ∂L = tril(∂factors[:, 1:q], -1)
-            end
-            L = UnitLowerTriangular(factors1)
-            U1 = UpperTriangular(factors1)
-            triu!(rmul!(∂A1, U1'))
-            ∂A1 .+= tril!(mul!(lmul!(L', ∂L), ∂A2, U2', -1, 1), -1)
-            rdiv!(∂A1, U1')
-            ldiv!(L', ∂A)
-        else  # tall A, system is [P1*A; P2*A] = [L1*U; L2*U]
-            tril!(copyto!(∂A, ∂factors), -1)
-            @views begin
-                factors1 = factors[1:q, :]
-                L2 = factors[(q + 1):end, :]
-                ∂A1 = ∂A[1:q, :]
-                ∂A2 = ∂A[(q + 1):end, :]
-                ∂U = triu(∂factors[1:q, :])
-            end
-            U = UpperTriangular(factors1)
-            L1 = UnitLowerTriangular(factors1)
-            tril!(lmul!(L1', ∂A1), -1)
-            ∂A1 .+= triu!(mul!(rmul!(∂U, U'), L2', ∂A2, -1, 1))
-            ldiv!(L1', ∂A1)
-            rdiv!(∂A, U')
-        end
-        if pivot isa LU_RowMaximum
-            ∂A = ∂A[invperm(F.p), :]
-        end
-        return NoTangent(), ∂A, NoTangent()
-    end
+    lu_pullback(ȳ) = _lu_pullback(ȳ, A, pivot, F)
```

Comment:
the body of the function does not change. The name changes from `lu_pullback` to `_lu_pullback` and it gets the extra arguments `A`, `pivot`, and `F`.

I guess we could close over `m`, `n` rather than `A` here, will change when doing the rest of review